components/tidb_query/src/rpn_expr/impl_math.rs

```diff
@@ -347,6 +351,26 @@ fn pow(lhs: &Option<Real>, rhs: &Option<Real>) -> Result<Option<Real>> {
     }
 }
 
+#[inline]
+#[rpn_fn(capture = [metadata], metadata_ctor = init_rng_data)]
+fn rand(metadata: &RefCell<MySQLRng>) -> Result<Option<Real>> {
```

Comment:
Actually I prefer to use a Thread Local Storage implementation for `rand()`, so that both vectorized and non-vectorized one can be benefited. What do you think? @andylokandy

Reply:
I don't think the ThreadLocal will work. The rng state should be mantained across batches, where the executor may be executed on different threads.

Reply:
There is already a rng cache mechanism in non-vectorized framework. What we have to do is replacing the XorRng with MySQLRng.

Reply:
Hmmmm I can't get your point. Why would you maintain a state for a non-seeded PRNG, considering that it will always generate a non-deterministic random number?

Reply:
oh, I thought it's rand with seed. Indeed we can cache it in LocalThread.

Reply:
Do you prefer the Thread Local approach? In this way, we can remove the rand state in non-vectorized function as well. The rand state introduces cost for ALL non-vectorized functions limited to the architecture and I dislike it very much..

Reply:
Do you mean we need to  use Thread Local to cache a `RefCell<MySQLRng>` and use it in rand()?
Or I have misunderstand?

Reply:
Yes, you're right.

Reply:
Ok, got it. Let me try tomorrow.

Reply:
You can also wrap your thread local logic in MySQLRng (or create a wrapper over it), so that duplicate code can be minimized.

Reply:
You mean use thread_local! in MySQLRng and use it in new()?

Reply:
> You can also wrap your thread local logic in MySQLRng (or create a wrapper over it), so that duplicate code can be minimized.

I think this involved unnecesary complexity into MySQLRng, since the rpn fn takes the responsibility of the caching behavior.

Reply:
Yes, and you can encapsulate it in a `ThreadLocalMySQLRng`.

Reply:
@andylokandy Yes, I agree with you. I think wrapping it in another structure is cleaner.

Reply:
Maybe we can use the bare ThreadLocal?

Reply:
PTAL~